src/expr/core/src/table_function/user_defined.rs

```diff
@@ -123,21 +124,48 @@ impl UserDefinedTableFunction {
 pub fn new_user_defined(prost: &PbTableFunction, chunk_size: usize) -> Result<BoxedTableFunction> {
     let udf = prost.get_udf()?;
 
-    let identifier = udf.get_identifier()?;
+    let arg_types = udf.arg_types.iter().map(|t| t.into()).collect::<Vec<_>>();
     let return_type = DataType::from(prost.get_return_type()?);
 
     let language = udf.language.as_str();
     let runtime = udf.runtime.as_deref();
     let link = udf.link.as_deref();
 
+    let name_in_runtime = if udf.version() < PbUdfExprVersion::NameInRuntime {
+        if language == "rust" || language == "wasm" {
+            // The `identifier` value of Rust and WASM UDF before `NameInRuntime`
+            // is not used any more. And unfortunately, we don't have the original name
+            // in `PbUserDefinedFunctionMetadata`, so we need to extract the name from
+            // the old `identifier` value (e.g. `foo()->int32`).
+            let old_identifier = udf
+                .identifier
+                .as_ref()
+                .expect("Rust/WASM UDF must have identifier");
+            Some(
+                old_identifier
+                    .split_once("(")
+                    .expect("the old identifier must contain `(`")
+                    .0,
+            )
+        } else {
+            // `identifier`s of other UDFs already mean `name_in_runtime` before `NameInRuntime`.
+            udf.identifier.as_deref()
+        }
+    } else {
+        // after `PbUdfExprVersion::NameInRuntime`, `identifier` means `name_in_runtime`
+        udf.identifier.as_deref()
+    }
+    .expect("SQL UDF won't get here, other UDFs must have `name_in_runtime`");
```

Comment:
Is it possible to make this an extension to the `PbUdf` struct, to clarify that the logic is intended for compatibility purposes?

Reply:
done